nixos/modules/hardware/device-tree.nix

```diff
@@ -0,0 +1,56 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.hardware.deviceTree;
+in {
+  options = {
+      hardware.deviceTree = {
+        enable = mkOption {
+          default = pkgs.stdenv.hostPlatform.platform.kernelDTB or false;
+          type = types.bool;
+          description = ''
+            Build device tree files. These are used to describe the
+            non-discoverable hardware of a system.
+          '';
+        };
+
+        base = mkOption {
+          default = "${config.boot.kernelPackages.kernel}/dtbs";
+          defaultText = "\${config.boot.kernelPackages.kernel}/dtbs";
+          example = literalExample "pkgs.deviceTree.raspberryPiDtbs";
+          type = types.nullOr types.path;
+          description = ''
+            The package containing the base device-tree (.dtb) to boot. Contains
+            device trees bundled with the Linux kernel by default.
+          '';
+        };
+
+        overlays = mkOption {
+          default = [];
+          example = literalExample
+            "[\"\${pkgs.deviceTree.raspberryPiOverlays}/w1-gpio.dtbo\"]";
+          type = types.listOf types.path;
+          description = ''
+            A path containing device tree overlays (.dtbo) to be applied to all
+            base device-trees.
+          '';
+        };
+
+        package = mkOption {
+          default = null;
+          type = types.nullOr types.path;
+          description = ''
+            A path containing device tree overlays (.dtbo) to be applied to all
+            base device-trees. Overrides `base` and `overlays`.
+          '';
+        };
+      };
+  };
+
+  config = mkIf (cfg.enable) {
+    hardware.deviceTree.package = if (cfg.overlays != [])
+      then pkgs.deviceTree.applyOverlays cfg.base cfg.overlays else cfg.base;
```

Comment:
This prevents you from setting package manually. I believe that if if `package` is settable as an option there probably needs to be some kind of logic like

```nix
mkIf (cfg.enable) 
  (mkIf (cfg.overlays != [] || cfg.base != null) { hardware.deviceTree.package = ...; })
```


I.e. With just `package` set I'm getting 

```
The unique option `hardware.deviceTree.package' is defined multiple times, in `...'  and `/nix/store/.../nixos/modules/hardware/device-tree.nix'
```

Reply:
I don't think `package` needs to bet settable - if `overlays` is empty, then `base` just becomes `package`. Is making `internal = true;` the correct way of making it not settable?

Reply:
> Is making internal = true; the correct way of making it not settable?

I'm not sure, possibly? Should it be exposed via `mkOption` at all? I imagined that it would have just been a simple let binding if it wasn't intended to be exposed as a settable option, but I could be wrong about that.

```nix
let 
  package = if (cfg.overlays != [])
       then pkgs.deviceTree.applyOverlays cfg.base cfg.overlays else cfg.base;
in
   ...
```